### PR TITLE
Development heartbeats

### PR DESCRIPTION
**Pull request by xmonader**
Updates for #298 

Creates heartbeat endpoint and support it in server and client and pushing it to influxdb

Related: https://github.com/zero-os/0-templates/pull/164

### CODE REVIEW

File: JumpScale9Lib/servers/grid_capacity/api_spec/main.raml [hunk 1]
```diff
@@ -134,6 +134,41 @@ types:
...
+        type: integer
+        description: uptime of the node in seconds
+        required: false
+    example:
+      node_id: 544546f60261
+      farmer_id: JohnDoe
+      os_version: "heads/v1.2.2 6b693a496de940b26ee1a2356b67f7d65767c13f"
+      robot_address: "http://10.241.100.10:6600"
+      uptime: 92682
+      location:
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/api_spec/main.raml]: example is not correct, there is no location in the type

File: JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+dir_path = os.path.dirname(os.path.realpath(__file__))
+Heartbeat_schema = JSON.load(open(dir_path + '/schema/Heartbeat_schema.json'))
+Heartbeat_schema_resolver = jsonschema.RefResolver('file://' + dir_path + '/schema/', Heartbeat_schema)
+Heartbeat_schema_validator = Draft4Validator(Heartbeat_schema, resolver=Heartbeat_schema_resolver)
+
+
+def RecordCapacityBeatHandler():
+    inputs = request.get_json()
+    # refresh jwt is needed otherwise return original
+    jwt = j.clients.itsyouonline.refresh_jwt_token(inputs.get('farmer_id'))
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py]: you can check if the jwt is expired first, so you don't always make a call to IYO

File: JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        return jsonify(errors="bad request body: {}".format(e)), 400
+
+
+    inputs.pop('farmer_id')
+
+    # heartbeat is subset of capacity.
+    inputs['farmer'] = iyo_organization
+    inputs['updated'] = datetime.now()
+
+    capacity = Capacity(**inputs)
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py]: Why do you create a capacity object here ?

File: JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py (hunk 2 above)
Comment (xmonader) [reply JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py]: Isn't it the case already? https://github.com/Jumpscale/lib9/blob/master/JumpScale9Lib/clients/itsyouonline/IYOFactory.py#L48

Comment (zaibon) [reply JumpScale9Lib/servers/grid_capacity/server/handlers/RecordCapacityBeatHandler.py]: indeed, didn't know this method does that, good :-)

File: JumpScale9Lib/servers/grid_capacity/server/app.py [hunk 4]
```diff
@@ -8,14 +8,23 @@
...
 app = Flask(__name__)
+
 app.secret_key = os.urandom(24)
 configure(app, settings.IYO_CLIENTID, settings.IYO_SECRET, settings.IYO_CALLBACK, '/callback', None, True, True, 'organization')
 
 # connect to mongodb
 j.clients.mongoengine.get('capacity', interactive=False)
 
+
+# from flask_debugtoolbar import DebugToolbarExtension
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/app.py]: clean this up ;-)

File: JumpScale9Lib/servers/grid_capacity/server/templates/capacity.html [hunk 5]
```diff
@@ -149,11 +149,13 @@ <h5 class="modal-title" id="nodeDetailModalLabel">Node detail</h5>
...
-            <p>Node ID: <code>{{node.node_id}}</code></p>
-            <p>Node 0-robot address: <code>{{node.robot_address}}</code></p>
-            <p>0-OS version: <code>{{node.os_version}}</code></p>
-            <p>Farm: <code><a href="/farmers#{{node.farmer.id}}">{{node.farmer.name}}</a></code></p>
-            <!-- <p>Uptime: <span class="badge badge-{{node.updated | deltatime_color}}">{{ node.uptime | uptime }}</span></p> -->
+            <p>Node ID: <code>{{node['node_id']}}</code></p>
+            <p>Node 0-robot address: <code>{{node['robot_address']}}</code></p>
+            <p>0-OS version: <code>{{node['os_version']}}</code></p>
+            <p>Farm: <code>{{node.farmer.name}}</code></p>
+            <p>Uptime: <span class="badge badge-{{node['updated'] | deltatime_color}}">{{ node['uptime'] | uptime }}</span></p>
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/templates/capacity.html]: we don't want to show uptime.

File: JumpScale9Lib/servers/grid_capacity/server/templates/capacity.html [hunk 6]
```diff
@@ -149,11 +149,13 @@ <h5 class="modal-title" id="nodeDetailModalLabel">Node detail</h5>
...
+            <p>Up time in last month (in hours): <code>{{ node | node_uptime_last_month_in_hours }}</code></p></o>
+            <p>Up time in last time (%): <code>{{ node | node_uptime_last_month_in_percent }}</code></p>
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/templates/capacity.html]: I don't understand this sentence: `uptime in last time` ? what do you mean ?

File: JumpScale9Lib/servers/grid_capacity/server/templates/farmers.html [hunk 7]
```diff
@@ -33,6 +33,8 @@ <h5 class="modal-title" id="farmDetailModalLabel-{{farmer.id}}">Farm detail</h5>
                             <p>Farm ID: <code>{{farmer.id}}</code></p>
                             <p>Farm name: <code>{{ farmer.name}}</code></p>
                             <p>IYO organization: <code>{{farmer.iyo_organization }}</code></p>
+                            <p>Up time in last month (in hours): <code>{{farmer | farmer_uptime_last_month_in_hours}}</code></p></o>
+                            <p>Up time in last time (%): <code>{{farmer | farmer_uptime_last_month_in_percent}}</code></p>
```
Comment (zaibon) [inline JumpScale9Lib/servers/grid_capacity/server/templates/farmers.html]: same

File: JumpScale9Lib/clients/zero_os/sal/Capacity/Capacity.py [hunk 8]
```diff
@@ -132,5 +132,36 @@ def update_reserved(self, vms, vdisks, gateways):
         )
 
         client = j.clients.grid_capacity.get(interactive=False)
+        client.api.RegisterCapacity(data)
```
Comment (zaibon) [inline JumpScale9Lib/clients/zero_os/sal/Capacity/Capacity.py]: this should not be there. this is the update_reserved method, so we don't do the registration here.

**Pull request closed without merging**